x/checkpointing/prepare/transactions.go

```diff
@@ -0,0 +1,105 @@
+package prepare
+
+import (
+	"errors"
+	"fmt"
+
+	abci "github.com/cometbft/cometbft/abci/types"
+)
+
+// PrepareProposalTxs is used as an intermediary storage for transactions when creating
+// a proposal for `PrepareProposal`.
+type PrepareProposalTxs struct {
+	// Transactions.
+	CheckpointTx []byte
+	OtherTxs     [][]byte
+
+	// Bytes.
+	// In general, there's no need to check for int64 overflow given that it would require
+	// exabytes of memory to hit the max int64 value in bytes.
+	MaxBytes  uint64
+	UsedBytes uint64
+}
+
+// NewPrepareProposalTxs returns a new `PrepareProposalTxs` given the request.
+func NewPrepareProposalTxs(
+	req *abci.RequestPrepareProposal,
+) (PrepareProposalTxs, error) {
+	if req.MaxTxBytes <= 0 {
+		return PrepareProposalTxs{}, errors.New("MaxTxBytes must be positive")
+	}
+
+	return PrepareProposalTxs{
+		MaxBytes:  uint64(req.MaxTxBytes),
+		UsedBytes: 0,
+	}, nil
+}
+
+// SetCheckpointTx sets the tx used for checkpoint
+func (t *PrepareProposalTxs) SetCheckpointTx(tx []byte) error {
+	oldBytes := uint64(len(t.CheckpointTx))
+	newBytes := uint64(len(tx))
+	if err := t.UpdateUsedBytes(oldBytes, newBytes); err != nil {
+		return err
+	}
+	t.CheckpointTx = tx
+	return nil
+}
+
+// AddOtherTxs adds txs to the "other" tx category and return early
+// if all the remaining bytes are used up
+func (t *PrepareProposalTxs) AddOtherTxs(allTxs [][]byte) error {
```

Comment:
Maybe `ReplaceOtherTxs` ? `Add` suggests that it would extend exsiting ones but in the first line we are doing 
```go
t.OtherTxs = make([][]byte, 0, len(allTxs))
```
which mean all other txs will be cleared